### PR TITLE
RskAddress class

### PR DESCRIPTION
**Pull request by tinchou**
This introduces an `RskAddress` class. Some of the benefits include:
* stop passing around `String`s and `byte[]`s
* ensure all addresses are valid (20 bytes long)
* type safety (e.g. you can't mix up parameters when passing multiple `byte[]`s)
* `equals` and `hashCode` allows using with standard collections

### CODE REVIEW

File: rskj-core/src/main/java/co/rsk/rpc/modules/personal/PersonalModuleWalletEnabled.java [hunk 1]
```diff
@@ -63,7 +64,7 @@ public void init(RskSystemProperties properties) {
         // wallets in production for security reasons.
         Account coinbaseAccount = properties.localCoinbaseAccount();
         if (coinbaseAccount != null) {
-            this.wallet.addAccount(coinbaseAccount);
+            this.wallet.addAccount(coinbaseAccount).getBytes();
```
Comment (diega) [inline rskj-core/src/main/java/co/rsk/rpc/modules/personal/PersonalModuleWalletEnabled.java]: I think this `.getBytes()` call should be removed

File: rskj-core/src/main/java/co/rsk/validators/BlockTxsValidationRule.java [hunk 2]
```diff
@@ -76,21 +76,20 @@ public boolean isValid(Block block, Block parent) {
...
             }
 
-            byte[] txSender = tx.getSender();
-            ByteArrayWrapper key = new ByteArrayWrapper(txSender);
-            BigInteger expectedNonce = curNonce.get(key);
+            RskAddress sender = tx.getSender();
+            BigInteger expectedNonce = curNonce.get(sender);
             if (expectedNonce == null) {
-                expectedNonce = parentRepo.getNonce(txSender);
+                expectedNonce = parentRepo.getNonce(sender.getBytes());
```
Comment (diega) [inline rskj-core/src/main/java/co/rsk/validators/BlockTxsValidationRule.java]: Maybe `org.ethereum.core.Repository` could be refactored as part of this work, it has a lot of `byte[]` referring RskAddresses

File: rskj-core/src/main/java/org/ethereum/core/PendingTransaction.java [hunk 3]
```diff
@@ -64,7 +64,7 @@ public long getBlockNumber() {
 
     public byte[] getSender() {
```
Comment (diega) [inline rskj-core/src/main/java/org/ethereum/core/PendingTransaction.java]: I think this method is not used anywhere, in that case the `byte[] sender` field can be removed

File: rskj-core/src/main/java/org/ethereum/core/Transaction.java [hunk 4]
```diff
@@ -615,4 +615,13 @@ public static Transaction create(byte[] nonce, Web3.CallArguments args){
 
         return new Transaction(nonce, hexArgs.getGasPrice(), hexArgs.getGasLimit(), hexArgs.getToAddress(), hexArgs.getValue(), hexArgs.getData());
     }
+
+    public static RskAddress parseRskAddress(@Nullable byte[] bytes) {
```
Comment (diega) [inline rskj-core/src/main/java/org/ethereum/core/Transaction.java]: doesn't make sense to make it part of the `RskAddress` constructor?

File: rskj-core/src/main/java/org/ethereum/vm/program/invoke/ProgramInvokeFactoryImpl.java [hunk 5]
```diff
@@ -51,15 +51,15 @@ public ProgramInvoke createProgramInvoke(Transaction tx, int txindex, Block bloc
...
 
         /***         ORIGIN op       ***/
         // YP: This is the sender of original transaction; it is never a contract.
-        byte[] origin = tx.getSender();
+        byte[] origin = tx.getSender().getBytes();
 
         /***         CALLER op       ***/
         // YP: This is the address of the account that is directly responsible for this execution.
-        byte[] caller = tx.getSender();
+        byte[] caller = tx.getSender().getBytes();
```
Comment (diega) [inline rskj-core/src/main/java/org/ethereum/vm/program/invoke/ProgramInvokeFactoryImpl.java]: For extra points you could refactor the `org.ethereum.vm.program.invoke.ProgramInvokeImpl` constructor as well :)

File: rskj-core/src/main/java/co/rsk/core/RskAddress.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+     * This is the size of an RSK address in bytes.
+     */
+    private static final int LENGTH_IN_BYTES = 20;
+
+    private static final RskAddress NULL_ADDRESS = new RskAddress(new byte[LENGTH_IN_BYTES]);
+
+    /**
+     * This compares using the lexicographical order of the sender unsigned bytes.
+     */
+    public static final Comparator<RskAddress> COMPARATOR = Comparator.comparing(
```
Comment (diega) [inline rskj-core/src/main/java/co/rsk/core/RskAddress.java]: Maybe this class could just implement `Comparable`

Comment (tinchou) [reply rskj-core/src/main/java/co/rsk/core/RskAddress.java]: We discussed this offline and didn't agree on this. My view is that addresses don't have a natural order (this implies implementing `Comparable`), but just an order for a specific feature (e.g. serializing always in the same order).
I think it's a minor issue, maybe we should leave this out for now.

File: rskj-core/src/main/java/co/rsk/rpc/modules/personal/PersonalModuleWalletEnabled.java (hunk 1 above)
Comment (tinchou) [reply rskj-core/src/main/java/co/rsk/rpc/modules/personal/PersonalModuleWalletEnabled.java]: :+1:

File: rskj-core/src/main/java/co/rsk/validators/BlockTxsValidationRule.java (hunk 2 above)
Comment (tinchou) [reply rskj-core/src/main/java/co/rsk/validators/BlockTxsValidationRule.java]: I noticed that too, but the number of changes was getting out of control. Just created #410 to track that.

File: rskj-core/src/main/java/org/ethereum/core/PendingTransaction.java (hunk 3 above)
Comment (tinchou) [reply rskj-core/src/main/java/org/ethereum/core/PendingTransaction.java]: Just removed the whole class!

File: rskj-core/src/main/java/org/ethereum/core/Transaction.java (hunk 4 above)
Comment (tinchou) [reply rskj-core/src/main/java/org/ethereum/core/Transaction.java]: I thought about it, but I want to make sure that anyone using this constructor is very conscious about what data they're passing. This is just a wrapper around **a valid address**. The `Transaction` class has the responsibility of sanitizing the input here, with the business rules it has defined.

File: rskj-core/src/main/java/org/ethereum/vm/program/invoke/ProgramInvokeFactoryImpl.java (hunk 5 above)
Comment (tinchou) [reply rskj-core/src/main/java/org/ethereum/vm/program/invoke/ProgramInvokeFactoryImpl.java]: :sos:

**Comment by tinchou** [comment]
@martinmedina created the constructor as suggested and created https://github.com/rsksmart/rskj/issues/411

**Pull request merged**